In [1]:
from neuron import *
from encodingsource import *
from hsgs import *
from classical_neuron import *
from classical_pso import *
from sf import *
simulator = Aer.get_backend('qasm_simulator')
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


/anaconda3/lib/python3.7/site-packages/requests/__init__.py:91: RequestsDependencyWarning: urllib3 (1.25.10) or chardet (3.0.4) doesn't match a supported version!
  RequestsDependencyWarning)


### Criando neuronio usando hsgs, sf, encoding e comparando com o resultado clássico


In [5]:
inputVector = [-1,-1,1,1]#[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]#[1, 1, 1, 1, -1, 1, -1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
weightVector = [1,1,-1,1] #[-1,1,-1,-1,1,1,1,1,1,1,1,1,1,1,1,1]#[1, -1, 1, 1, -1, 1, 1, -1, -1, 1, 1, 1, 1, -1, 1, 1]

inputVector = [-1, 1, 1, 1, -1, 1, 1, -1, 1, -1, -1, -1, 1, 1, -1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, 1, -1, 1, -1, -1, -1, 1, 1, -1, -1, -1, -1, -1, -1, 1, -1, 1, 1, -1, -1, -1, 1, -1, 1, -1, -1, -1, -1, -1, 1, -1, 1, -1, 1, 1, -1, 1, 1, -1]
weightVector = [-0.03216409993230598, -0.1520999574184498, 0.11549307734856797, 0.07819667776917735, -0.08683451299393112, 0.07428209202455867, -0.011628350965060036, -0.058852035623669065, 0.02091024109530386, 0.18422984345689053, -0.12001327917718091, -0.17281808213991004, 0.004445115622186578, 0.09949334308845695, 0.0644906589346581, 0.17559849756856852, -0.088707282935652, -0.017908127406289862, -0.1026406028784107, 0.1885141061843412, 0.04853405915763267, -0.1894412214705037, -0.18971428207273894, -0.16080045829897144, -0.14494276618814597, 0.16133137423246743, 0.032410503109816365, 0.1554236889112614, -0.18999964114222914, -0.02773208631445002, -0.1769047447607832, 0.07340800548043382, 0.04977970093090443, -0.18319674721020804, -0.1611156274775103, -0.08041675891466839, -0.10661038807499998, -0.15819257805946796, -0.07129341872229379, 0.17219959076743557, -0.012891323765141861, 0.13884392920174257, -0.03493177735039414, -0.16967932466594843, -0.1344533324680956, 0.07710909951302578, 0.16230727329009179, -0.11811109546183819, 0.05762967585657125, 0.15720292553372192, -0.16562784891930163, -0.07348394951216673, -0.1298264325738957, -0.06989954242058431, 0.1662649564426879, 0.15549686702988463, -0.1345078290406552, 0.07132344671653838, 0.13820247035438069, -0.06455638157173038, 0.1549689794465881, -0.18061152100490427, 0.021178877641755214, -0.15711833018132063]

operator = "encoding-input"
neuron = createNeuron( inputVector, weightVector, operator)
resultadoEncoding = executeNeuron(neuron, simulator, threshold=None)
print("encoding-input", resultadoEncoding)   
operator = "encoding-weight"
neuron = createNeuron( inputVector, weightVector, operator)
resultadoEncoding = executeNeuron(neuron, simulator, threshold=None)
print("encoding-weight", resultadoEncoding)   
operator = "hsgs"
neuron = createNeuron(inputVector, weightVector, operator)
resultadoHSGS = executeNeuron(neuron, simulator, threshold=None)
print("hsgs", resultadoHSGS)

resultadoClassico = runClassicalNeuronReturnProbability(inputVector, weightVector)
print("classico", resultadoClassico)  

encoding-input 0.0465087890625
encoding-weight 0.21630859375
hsgs 0.25537109375
classico 0.21734167312340363


## Treinando Base simples

### Base com X, Quadrado e Círculo

In [113]:
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from sklearn import datasets
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split


X1 = [-1,  1,  1, -1,
       1, -1, -1,  1,
       1, -1, -1,  1,
      -1,  1,  1, -1]

Q1 = [-1, -1, -1, -1,
      -1,  1,  1, -1,
      -1,  1,  1, -1,
      -1, -1, -1, -1]

C1 = [ 1, -1, -1,  1,
      -1,  1,  1, -1,
      -1,  1,  1, -1,
       1, -1, -1,  1]

X2 = [-1,  1, -1, -1,
       1, -1, -1,  1,
       1, -1, -1,  1,
      -1,  1,  1, -1]
X3 = [-1,  1,  1, -1,
       1, -1, -1,  1,
       1, -1, -1, -1,
      -1,  1,  1, -1]

Q2 = [-1, -1, -1, -1,
      -1, -1,  1, -1,
      -1,  1,  1, -1,
      -1, -1, -1, -1]
Q3 = [-1, -1, -1, -1,
      -1,  1,  1, -1,
      -1, -1,  1, -1,
      -1, -1, -1, -1]
C2 = [ 1, -1, -1, -1,
      -1,  1,  1, -1,
      -1,  1,  1, -1,
       1, -1, -1,  1]
C3 = [ 1,  1, -1,  1,
      -1,  1,  1, -1,
      -1,  1,  1, -1,
       1, -1, -1,  1]

#X = [[-1,-1,-1,-1], [1,-1,-1,-1],[1,1,1,-1], [1,1,-1,-1]]
#y = [0,0,1,1]

#Xs_train, Xs_test, ys_train, ys_test = train_test_split(X, y, test_size=0.5, stratify=y,  random_state = 123)

Xs_train = [X1, Q1, C1]
ys_train = [0, 1, 2]

Xs_test = [X2, X3, Q2, Q3, C2, C3]
ys_test = [0, 0, 1, 1, 2, 2]
print(Xs_train)
print(ys_train)
print(Xs_test)
print(ys_test)



[[-1, 1, 1, -1, 1, -1, -1, 1, 1, -1, -1, 1, -1, 1, 1, -1], [-1, -1, -1, -1, -1, 1, 1, -1, -1, 1, 1, -1, -1, -1, -1, -1], [1, -1, -1, 1, -1, 1, 1, -1, -1, 1, 1, -1, 1, -1, -1, 1]]
[0, 1, 2]
[[-1, 1, -1, -1, 1, -1, -1, 1, 1, -1, -1, 1, -1, 1, 1, -1], [-1, 1, 1, -1, 1, -1, -1, 1, 1, -1, -1, -1, -1, 1, 1, -1], [-1, -1, -1, -1, -1, -1, 1, -1, -1, 1, 1, -1, -1, -1, -1, -1], [-1, -1, -1, -1, -1, 1, 1, -1, -1, -1, 1, -1, -1, -1, -1, -1], [1, -1, -1, -1, -1, 1, 1, -1, -1, 1, 1, -1, 1, -1, -1, 1], [1, 1, -1, 1, -1, 1, 1, -1, -1, 1, 1, -1, 1, -1, -1, 1]]
[0, 0, 1, 1, 2, 2]


## Treinamento

In [122]:
thresholdParameter = 0.5 #nao é usado

def treinamentoNeuronio(operator,inputVector, weightVector, y_train):
    if (operator == "hsgs"):
        wBinaryBinary = deterministicBinarization(weightVector) # Binarization of Real weights
        neuron = createNeuron(inputVector, wBinaryBinary, operator)
        resultado = executeNeuron(neuron, simulator, threshold=None)
        deltaRule(inputVector, weightVector, threshold=thresholdParameter, lr=lrParameter, y_train=y_train, out=resultado)
        return resultado
    elif (operator == "encoding-weight"):
        #wBinaryBinary = deterministicBinarization(weightVector) # Binarization of Real weights
        neuron = createNeuron(inputVector, weightVector, operator)
        resultado = executeNeuron(neuron, simulator, threshold=None)
        deltaRule(inputVector, weightVector, threshold=thresholdParameter, lr=lrParameter, y_train=y_train, out=resultado)
        return resultado
        
nClasses = 3
weightVectorsHSGS = []#[[-1, -1, -1, -1], [1, 1, 1, -1]]#[[-1, -1, -1, -1], [-1, -1, -1, -1]]#[[1,1,1,1]]*nClasses
weightVectorsEncoding = []#[[-1, -1, -1, -1], [1, 1, 1, -1]]#[[-1, -1, -1, -1], [-1, -1, -1, -1]]#[[1,1,1,1]]*nClasses

input_dim = 2**4

for i in range(nClasses):
    weightVectorsHSGS.append(deterministicBinarization(np.random.uniform(-1, 1, input_dim)))
    weightVectorsEncoding.append(deterministicBinarization(np.random.uniform(-1, 1, input_dim)))

limiarErroToleravel = 0.03
n_epochs = 200

for lrParameter in [0.2]:
    print("taxa de aprendizado",lrParameter )
    tamTreinamento = len(Xs_train)

    for iteration in range(n_epochs):
        erroHSGS = 0
        erroEncoding = 0

        errosHSGS=[]
        errosEncoding = []
        for posicaoTreinamento in range(tamTreinamento):
            
            inputVector = Xs_train[posicaoTreinamento] # inputVectors[posicaoTreinamento]
            y_train = ys_train[posicaoTreinamento]

            #print(inputVector, y_train)

            """
            executando o HSGS
            """
            operator = "hsgs"
            resultadoHSGS = treinamentoNeuronio(operator,inputVector, weightVectorsHSGS[y_train], y_train=1)
            """
            executando o encoding
            """
            operator = "encoding-weight"
            resultadoEncoding = treinamentoNeuronio(operator,inputVector, weightVectorsEncoding[y_train], y_train=1)

            #print("PESOS ENCODING",weightVectorsEncoding)


            #print("resultadoHSGS", resultadoHSGS)
            #print("encoding", resultadoEncoding)
            """
            erros
            """
            """
            #old
            resultadoEncoding_bin = 0
            if abs(resultadoEncoding) > threshold:
                resultadoEncoding_bin = 1

            resultadoHSGS_bin = 0
            if abs(resultadoHSGS) > threshold:
                resultadoHSGS_bin = 1


            erroHSGS += abs(resultadoHSGS_bin)####abs(resultadoHSGS_bin-y_train)
            erroEncoding += abs(resultadoEncoding_bin)####abs(resultadoEncoding_bin-y_train)
            """
            errosHSGS.append(1-resultadoHSGS)
            errosEncoding.append(1-resultadoEncoding)

            erroHSGS += abs(1-resultadoHSGS)####abs(resultadoHSGS_bin-y_train)
            erroEncoding += abs(1-resultadoEncoding)####abs(resultadoEncoding_bin-y_train)
            
        print("erro HSGS", erroHSGS, "erro encoding", erroEncoding)
        #print("erros", errosHSGS, errosEncoding)
        #if erroEncoding < limiarErroToleravel or erroHSGS < limiarErroToleravel:
        #        break
    print("erros", errosHSGS, errosEncoding)


taxa de aprendizado 0.2
erro HSGS 2.9263916015625 erro encoding 2.9049072265625
erro HSGS 2.9197998046875 erro encoding 2.6484375
erro HSGS 2.9212646484375 erro encoding 2.3145751953125
erro HSGS 2.9268798828125 erro encoding 2.0047607421875
erro HSGS 2.9232177734375 erro encoding 1.7681884765625
erro HSGS 2.9190673828125 erro encoding 1.5526123046875
erro HSGS 0.0 erro encoding 1.4083251953125
erro HSGS 0.0 erro encoding 1.275390625
erro HSGS 0.0 erro encoding 1.1689453125
erro HSGS 0.0 erro encoding 1.0924072265625
erro HSGS 0.0 erro encoding 1.0023193359375
erro HSGS 0.0 erro encoding 0.9619140625
erro HSGS 0.0 erro encoding 0.8973388671875
erro HSGS 0.0 erro encoding 0.838134765625
erro HSGS 0.0 erro encoding 0.7890625
erro HSGS 0.0 erro encoding 0.77392578125
erro HSGS 0.0 erro encoding 0.7257080078125
erro HSGS 0.0 erro encoding 0.7049560546875
erro HSGS 0.0 erro encoding 0.6729736328125
erro HSGS 0.0 erro encoding 0.65234375
erro HSGS 0.0 erro encoding 0.630126953125
erro HSGS 0

erro HSGS 0.0 erro encoding 0.12744140625
erro HSGS 0.0 erro encoding 0.1253662109375
erro HSGS 0.0 erro encoding 0.127197265625
erro HSGS 0.0 erro encoding 0.1295166015625
erro HSGS 0.0 erro encoding 0.123291015625
erro HSGS 0.0 erro encoding 0.12841796875
erro HSGS 0.0 erro encoding 0.1265869140625
erro HSGS 0.0 erro encoding 0.129150390625
erro HSGS 0.0 erro encoding 0.12841796875
erro HSGS 0.0 erro encoding 0.12353515625
erro HSGS 0.0 erro encoding 0.1292724609375
erro HSGS 0.0 erro encoding 0.1273193359375
erros [0.0, 0.0, 0.0] [0.0439453125, 0.0423583984375, 0.041015625]


## Fazendo teste

In [130]:
print("PESOS HSGS", weightVectorsHSGS)
print()
print("PESOS Encoding", weightVectorsEncoding)
print()

errosHSGS = []
errosEncoding = []
for i in range(20):
    erroHSGS = 0
    erroEncoding = 0

    for posicaoTreinamento in range(len(Xs_test)):
        inputVector = Xs_test[posicaoTreinamento] # inputVectors[posicaoTreinamento]
        y_train = ys_test[posicaoTreinamento]

        valorMaiorHSGS=0
        neuronMaiorHSGS=0

        valorMaiorEncoding=0
        neuronMaiorEncoding=0

        for neuronClass in range(nClasses):
            operator = "hsgs"
            wBinaryBinary = deterministicBinarization(weightVectorsHSGS[neuronClass]) # Binarization of Real weights
            neuron = createNeuron(inputVector, wBinaryBinary, operator)
            resultadoHSGS1 = executeNeuron(neuron, simulator, threshold=None)

            if(resultadoHSGS1>valorMaiorHSGS):
                neuronMaiorHSGS = neuronClass
                valorMaiorHSGS = resultadoHSGS1

            operator = "encoding-weight"
            #wBinaryBinary = deterministicBinarization(weightVectorsEncoding[neuronClass]) # Binarization of Real weights
            neuron = createNeuron(inputVector, weightVectorsEncoding[neuronClass], operator)
            resultadoEncoding1 = executeNeuron(neuron, simulator, threshold=None)

            if(resultadoEncoding1 > valorMaiorEncoding):
                neuronMaiorEncoding = neuronClass
                valorMaiorEncoding = resultadoEncoding1


        """
        erros
        """
        erroHSGS_bin = 0
        if (neuronMaiorHSGS != y_train):   
            erroHSGS_bin = 1

        erroEncoding_bin = 0
        if (neuronMaiorEncoding != y_train):   
            erroEncoding_bin = 1


        #print("\nerro Encoding:", abs(resultadoEncoding_bin-y_train))
        #print("erro HSGS:", abs(resultadoHSGS_bin-y_train))
        erroHSGS += erroHSGS_bin####abs(resultadoHSGS_bin-y_train)
        erroEncoding += erroEncoding_bin####abs(resultadoEncoding_bin-y_train)
    print()
    print("erro HSGS", erroHSGS, "erro encoding", erroEncoding)
    errosHSGS.append(erroHSGS)
    errosEncoding.append(erroEncoding)

print("ERROS HSGS", errosHSGS, np.average(errosHSGS))
print("ERROS ENCODING", errosEncoding, np.average(errosEncoding))



PESOS HSGS [[0.19999999999999996, -2.1999999999999997, -2.1999999999999997, 2.1999999999999997, -0.19999999999999996, 0.19999999999999996, 2.1999999999999997, -0.19999999999999996, -0.19999999999999996, 0.19999999999999996, 2.1999999999999997, -0.19999999999999996, 2.1999999999999997, -2.1999999999999997, -0.19999999999999996, 2.1999999999999997], [2.1814697265625003, 0.18146972656250007, 0.18146972656250007, 0.18146972656250007, 2.1814697265625003, -0.18146972656250007, -0.18146972656250007, 0.18146972656250007, 2.1814697265625003, -2.1814697265625003, -0.18146972656250007, 2.1814697265625003, 0.18146972656250007, 2.1814697265625003, 2.1814697265625003, 0.18146972656250007], [-0.12585449218750003, 2.1258544921875004, 2.1258544921875004, -2.1258544921875004, 2.1258544921875004, -2.1258544921875004, -2.1258544921875004, 2.1258544921875004, 0.12585449218750003, -2.1258544921875004, -0.12585449218750003, 0.12585449218750003, -0.12585449218750003, 2.1258544921875004, 0.12585449218750003, -

In [131]:
taxasHSGS = [0, 1, 2, 0, 1, 2, 3, 2, 1, 0, 2, 4, 0, 3, 2, 1, 2, 2, 1, 2]
mediaHSGS = 0
for t in taxasHSGS:
    mediaHSGS += t/4
print(mediaHSGS/20)

taxasEncoding = [0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1]
mediaEncoding = 0
for t in taxasEncoding:
    mediaEncoding += t/4
print(mediaEncoding/20)

0.3875
0.15


## Resultados anteriores:

[[0.19999999999999996, -2.1999999999999997, -2.1999999999999997, 2.1999999999999997, -0.19999999999999996, 0.19999999999999996, 2.1999999999999997, -0.19999999999999996, -0.19999999999999996, 0.19999999999999996, 2.1999999999999997, -0.19999999999999996, 2.1999999999999997, -2.1999999999999997, -0.19999999999999996, 2.1999999999999997], [2.1814697265625003, 0.18146972656250007, 0.18146972656250007, 0.18146972656250007, 2.1814697265625003, -0.18146972656250007, -0.18146972656250007, 0.18146972656250007, 2.1814697265625003, -2.1814697265625003, -0.18146972656250007, 2.1814697265625003, 0.18146972656250007, 2.1814697265625003, 2.1814697265625003, 0.18146972656250007], [-0.12585449218750003, 2.1258544921875004, 2.1258544921875004, -2.1258544921875004, 2.1258544921875004, -2.1258544921875004, -2.1258544921875004, 2.1258544921875004, 0.12585449218750003, -2.1258544921875004, -0.12585449218750003, 0.12585449218750003, -0.12585449218750003, 2.1258544921875004, 0.12585449218750003, -2.1258544921875004]]
[[5.598706054687503, -5.598706054687503, -3.5987060546874985, 5.598706054687503, -5.598706054687503, 5.598706054687503, 3.5987060546874985, -3.5987060546874985, -3.5987060546874985, 5.598706054687503, 5.598706054687503, -5.598706054687503, 3.5987060546874985, -5.598706054687503, -3.5987060546874985, 3.5987060546874985], [5.598754882812498, 3.598754882812499, 5.598754882812498, 5.598754882812498, 3.598754882812499, -3.598754882812499, -5.598754882812498, 5.598754882812498, 5.598754882812498, -5.598754882812498, -3.598754882812499, 3.598754882812499, 5.598754882812498, 3.598754882812499, 3.598754882812499, 5.598754882812498], [-3.4143066406250004, 3.4143066406250004, 5.4143066406249964, -3.4143066406250004, 5.4143066406249964, -5.4143066406249964, -3.4143066406250004, 3.4143066406250004, 5.4143066406249964, -5.4143066406249964, -5.4143066406249964, 5.4143066406249964, -5.4143066406249964, 5.4143066406249964, 3.4143066406250004, -5.4143066406249964]]
classe esperada 0
NEURONIO 0
res HSGS 0.767333984375 resultadoEncoding1 0.7879638671875
NEURONIO 1
res HSGS 0.13330078125 resultadoEncoding1 0.0633544921875
NEURONIO 2
res HSGS 0.7650146484375 resultadoEncoding1 0.7042236328125
classe esperada 0
NEURONIO 0
res HSGS 0.761474609375 resultadoEncoding1 0.697265625
NEURONIO 1
res HSGS 0.1405029296875 resultadoEncoding1 0.0943603515625
NEURONIO 2
res HSGS 0.7738037109375 resultadoEncoding1 0.69091796875
classe esperada 1
NEURONIO 0
res HSGS 0.140625 resultadoEncoding1 0.1304931640625
NEURONIO 1
res HSGS 0.76416015625 resultadoEncoding1 0.7872314453125
NEURONIO 2
res HSGS 0.1427001953125 resultadoEncoding1 0.1392822265625
classe esperada 1
NEURONIO 0
res HSGS 0.146240234375 resultadoEncoding1 0.1322021484375
NEURONIO 1
res HSGS 0.7625732421875 resultadoEncoding1 0.6927490234375
NEURONIO 2
res HSGS 0.1474609375 resultadoEncoding1 0.135498046875
classe esperada 2
NEURONIO 0
res HSGS 0.7635498046875 resultadoEncoding1 0.6953125
NEURONIO 1
res HSGS 0.38720703125 resultadoEncoding1 0.2991943359375
NEURONIO 2
res HSGS 0.7681884765625 resultadoEncoding1 0.79248046875
classe esperada 2
NEURONIO 0
res HSGS 0.76123046875 resultadoEncoding1 0.6961669921875
NEURONIO 1
res HSGS 0.1470947265625 resultadoEncoding1 0.094970703125
NEURONIO 2
res HSGS 0.77587890625 resultadoEncoding1 0.7967529296875

erro HSGS 1 erro encoding 0

[[0.19999999999999996, -2.1999999999999997, -2.1999999999999997, 2.1999999999999997, -0.19999999999999996, 0.19999999999999996, 2.1999999999999997, -0.19999999999999996, -0.19999999999999996, 0.19999999999999996, 2.1999999999999997, -0.19999999999999996, 2.1999999999999997, -2.1999999999999997, -0.19999999999999996, 2.1999999999999997], [2.1814697265625003, 0.18146972656250007, 0.18146972656250007, 0.18146972656250007, 2.1814697265625003, -0.18146972656250007, -0.18146972656250007, 0.18146972656250007, 2.1814697265625003, -2.1814697265625003, -0.18146972656250007, 2.1814697265625003, 0.18146972656250007, 2.1814697265625003, 2.1814697265625003, 0.18146972656250007], [-0.12585449218750003, 2.1258544921875004, 2.1258544921875004, -2.1258544921875004, 2.1258544921875004, -2.1258544921875004, -2.1258544921875004, 2.1258544921875004, 0.12585449218750003, -2.1258544921875004, -0.12585449218750003, 0.12585449218750003, -0.12585449218750003, 2.1258544921875004, 0.12585449218750003, -2.1258544921875004]]
[[5.598706054687503, -5.598706054687503, -3.5987060546874985, 5.598706054687503, -5.598706054687503, 5.598706054687503, 3.5987060546874985, -3.5987060546874985, -3.5987060546874985, 5.598706054687503, 5.598706054687503, -5.598706054687503, 3.5987060546874985, -5.598706054687503, -3.5987060546874985, 3.5987060546874985], [5.598754882812498, 3.598754882812499, 5.598754882812498, 5.598754882812498, 3.598754882812499, -3.598754882812499, -5.598754882812498, 5.598754882812498, 5.598754882812498, -5.598754882812498, -3.598754882812499, 3.598754882812499, 5.598754882812498, 3.598754882812499, 3.598754882812499, 5.598754882812498], [-3.4143066406250004, 3.4143066406250004, 5.4143066406249964, -3.4143066406250004, 5.4143066406249964, -5.4143066406249964, -3.4143066406250004, 3.4143066406250004, 5.4143066406249964, -5.4143066406249964, -5.4143066406249964, 5.4143066406249964, -5.4143066406249964, 5.4143066406249964, 3.4143066406250004, -5.4143066406249964]]
classe esperada 0
NEURONIO 0
res HSGS 0.7655029296875 resultadoEncoding1 0.7860107421875
NEURONIO 1
res HSGS 0.1431884765625 resultadoEncoding1 0.0677490234375
NEURONIO 2
res HSGS 0.76416015625 resultadoEncoding1 0.70361328125
classe esperada 0
NEURONIO 0
res HSGS 0.7689208984375 resultadoEncoding1 0.685791015625
NEURONIO 1
res HSGS 0.14306640625 resultadoEncoding1 0.0906982421875
NEURONIO 2
res HSGS 0.771240234375 resultadoEncoding1 0.689208984375
classe esperada 1
NEURONIO 0
res HSGS 0.1376953125 resultadoEncoding1 0.126220703125
NEURONIO 1
res HSGS 0.76416015625 resultadoEncoding1 0.782470703125
NEURONIO 2
res HSGS 0.1436767578125 resultadoEncoding1 0.1474609375
classe esperada 1
NEURONIO 0
res HSGS 0.1380615234375 resultadoEncoding1 0.1304931640625
NEURONIO 1
res HSGS 0.7630615234375 resultadoEncoding1 0.698974609375
NEURONIO 2
res HSGS 0.142822265625 resultadoEncoding1 0.1348876953125
classe esperada 2
NEURONIO 0
res HSGS 0.7633056640625 resultadoEncoding1 0.6915283203125
NEURONIO 1
res HSGS 0.3935546875 resultadoEncoding1 0.291015625
NEURONIO 2
res HSGS 0.7618408203125 resultadoEncoding1 0.7860107421875
classe esperada 2
NEURONIO 0
res HSGS 0.776611328125 resultadoEncoding1 0.7030029296875
NEURONIO 1
res HSGS 0.143310546875 resultadoEncoding1 0.093505859375
NEURONIO 2
res HSGS 0.7664794921875 resultadoEncoding1 0.7965087890625

erro HSGS 3 erro encoding 1

In [129]:
print("PESOS HSGS", weightVectorsHSGS)
print()
print("PESOS Encoding", weightVectorsEncoding)
print()

errosHSGS = []
errosEncoding = []
for i in range(20):
    erroHSGS = 0
    erroEncoding = 0

    limiar=0.5

    

    
    for posicaoTreinamento in range(len(Xs_test)):
        inputVector = Xs_test[posicaoTreinamento] # inputVectors[posicaoTreinamento]
        y_train = ys_test[posicaoTreinamento]

        valorMaiorHSGS=0
        neuronMaiorHSGS=0

        valorMaiorEncoding=0
        neuronMaiorEncoding=0

        for neuronClass in range(nClasses):
            operator = "hsgs"
            wBinaryBinary = deterministicBinarization(weightVectorsHSGS[neuronClass]) # Binarization of Real weights
            neuron = createNeuron(inputVector, wBinaryBinary, operator)
            resultadoHSGS1 = executeNeuron(neuron, simulator, threshold=None)

            if(resultadoHSGS1>valorMaiorHSGS):
                neuronMaiorHSGS = neuronClass
                valorMaiorHSGS = resultadoHSGS1

            operator = "encoding-weight"
            #wBinaryBinary = deterministicBinarization(weightVectorsEncoding[neuronClass]) # Binarization of Real weights
            neuron = createNeuron(inputVector, weightVectorsEncoding[neuronClass], operator)
            resultadoEncoding1 = executeNeuron(neuron, simulator, threshold=None)

            if(resultadoEncoding1 > valorMaiorEncoding):
                neuronMaiorEncoding = neuronClass
                valorMaiorEncoding = resultadoEncoding1


        """
        erros
        """
        erroHSGS_bin = 0
        if (neuronMaiorHSGS != y_train):   
            erroHSGS_bin = 1

        erroEncoding_bin = 0
        if (neuronMaiorEncoding != y_train):   
            erroEncoding_bin = 1


        #print("\nerro Encoding:", abs(resultadoEncoding_bin-y_train))
        #print("erro HSGS:", abs(resultadoHSGS_bin-y_train))
        erroHSGS += erroHSGS_bin####abs(resultadoHSGS_bin-y_train)
        erroEncoding += erroEncoding_bin####abs(resultadoEncoding_bin-y_train)
    print()
    print("erro HSGS", erroHSGS, "erro encoding", erroEncoding)
    errosHSGS.append(erroHSGS)
    errosEncoding.append(erroEncoding)

print("ERROS HSGS", errosHSGS, np.average(errosHSGS))
print("ERROS ENCODING", errosEncoding, np.average(errosEncoding))


PESOS HSGS [[0.19999999999999996, -2.1999999999999997, -2.1999999999999997, 2.1999999999999997, -0.19999999999999996, 0.19999999999999996, 2.1999999999999997, -0.19999999999999996, -0.19999999999999996, 0.19999999999999996, 2.1999999999999997, -0.19999999999999996, 2.1999999999999997, -2.1999999999999997, -0.19999999999999996, 2.1999999999999997], [2.1814697265625003, 0.18146972656250007, 0.18146972656250007, 0.18146972656250007, 2.1814697265625003, -0.18146972656250007, -0.18146972656250007, 0.18146972656250007, 2.1814697265625003, -2.1814697265625003, -0.18146972656250007, 2.1814697265625003, 0.18146972656250007, 2.1814697265625003, 2.1814697265625003, 0.18146972656250007], [-0.12585449218750003, 2.1258544921875004, 2.1258544921875004, -2.1258544921875004, 2.1258544921875004, -2.1258544921875004, -2.1258544921875004, 2.1258544921875004, 0.12585449218750003, -2.1258544921875004, -0.12585449218750003, 0.12585449218750003, -0.12585449218750003, 2.1258544921875004, 0.12585449218750003, -